pages/api/gifts/[id].ts

```diff
@@ -0,0 +1,91 @@
+import axios, { isAxiosError } from 'axios';
+import { NextApiRequest, NextApiResponse } from 'next';
+import { Gift } from '~/pages';
+
+const baseURL = 'http://localhost:3001/gifts';
+
+const HANDLERS: Record<
+  string,
+  (req: NextApiRequest, res: NextApiResponse) => Promise<void>
+> = {
+  GET: handleGET,
+  POST: handlePOST,
+  PATCH: handlePATCH,
+  PUT: handlePUT,
+  DELETE: handleDELETE,
+} as const;
+
+export default async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse,
+) {
+  try {
+    const reqHandler = req.method !== undefined && HANDLERS[req.method];
+    if (reqHandler) {
+      await reqHandler(req, res);
+    } else {
+      return res
+        .status(405)
+        .send(
+          `${req.method} is not a valid method. Only GET requests are valid!`,
+        );
+    }
+  } catch (e) {
+    return errorFound(res, e);
+  }
+}
+
+async function handleGET(req: NextApiRequest, res: NextApiResponse) {
+  if (typeof req.query.id === 'string') {
+    const giftRequest = await axios.get(`${baseURL}/${req.query.id}`);
+    return res.status(giftRequest.status).send(giftRequest.data as Gift[]);
+  }
+  throw new Error('Invalid ID', { cause: 'idError' });
```

Comment:
Kun tätä if-statementtia ei ole, heittää ESLint tällaisen errorin `Invalid type "string | string[] | undefined" of template literal expression.` `req.query.id`:lle.

Tuohon on varmasti miljoona eri tapaa tehdä, tein sen nyt näin. `req.query.id` ei pitäisi koskaan voida olla `undefined`, eikä myöskään tyyppi `string[]`l, joten voiko tällaisissa tapauksissa tehdä näin: 

``const giftRequest = await axios.get(`${baseURL}/${req.query.id as string}`)``

Vai onko suotavaa tehdä aina jonkinlainen tyypin tarkistus?

Itse olen vähäisellä kokemuksellani sen kannalla, että tehdään aina tarkistus, jotta vältytään "vaarallisilta" ongelmilta. Mutta luulen että osaat sanoa paremmin :smile:

Reply:
Tässä on hyvä esimerkki tyyppien hyödyllisyydestä: kokeilematta sanoisin, että `req.query.id`voi olla undefined (esim. `?id=` ) tai string[] (esim. `?id=abc,def`.

Joka tapauksessa on paljon parempi tehdä tarkistus kuten olet tehnyt kuin käyttää `as`.

Ainoa muutos mitä tekisin niin yleensä itse teen näin päin tuon ehtolausekkeen. Selkeämpi etenkin jos on pidempi funktio:

```ts
  if (typeof req.query.id !== 'string')
    throw new Error('Invalid ID', { cause: 'idError' });

  const giftRequest = await axios.get(`${baseURL}/${req.query.id}`);
  return res.status(giftRequest.status).send(giftRequest.data as Gift[]);
```